### PR TITLE
Add support for reindex over CCS

### LINKED ISSUE #89120: Support reindex over CCS
### Description

Given we have registered remotes and are able to cross cluster search/replicate I would like to be able to reindex a remote index on the local cluster.

The normal index from remote requires a host to be set. The proxy endpoint used for the remote doesn't work as that's over transport and index from remote uses the http/REST interface. Using the ES rest requires a user to be created with the correct permission, effectively not using the registered remote.

Doing a local index results in index not found:

```
# normal search works
POST remote1:example-index/_search

# results in index not found
POST _reindex
{
  "source": {
    "index": "remote1:example-index"
  },
  "dest": {
    "index": "my-new-example-index"
  }
}
```

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "index_not_found_exception",
        "reason" : "no such index [remote1:example-index]",
        "resource.type" : "index_or_alias",
        "resource.id" : "remote1:example-index",
        "index_uuid" : "_na_",
        "index" : "remote1:example-index"
      }
    ],
    "type" : "index_not_found_exception",
    "reason" : "no such index [remote1:example-index]",
    "resource.type" : "index_or_alias",
    "resource.id" : "remote1:example-index",
    "index_uuid" : "_na_",
    "index" : "remote1:example-index"
  },
  "status" : 404
}
```

### PR DESCRIPTION
**Pull request by volodk85**
Reindexing requires `ignore_unavailable` search option to be set on request to execute reindex from remote - to be able to reference remote index in a format **cluster_name:index_name**
https://github.com/elastic/elasticsearch/blob/2e6834ad1103f7eab2e5a9cf2dcb5816ec4d2d17/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L1612-L1622

Unfortunately [Reindex API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#docs-reindex-api-query-params) does not have `ignore_unavailable` among supported query parameters. In the same time there is cluster level settings `cluster.remote.<cluster_name>.skip_unavailable` which allows check if remote cluster is configured to skip unavailable indices.

This PR checks if CCS/CCR is enabled and modifies request object if there is `skip_unavailable` flag is present.
Both re-index request formats are supported now

Request 1:
```
curl -X POST "localhost:19200/_reindex?pretty" -H 'Content-Type: application/json' -d'
{
  "source": {
        "remote": {
                "host": "http://172.18.0.2:9200"
        },
    "index": "my-index-000001"
  },
  "dest": {
    "index": "my-index-000001"
  }
}
'
```

Request 2:
```
curl -X POST "localhost:19200/_reindex?pretty" -H 'Content-Type: application/json' -d'
{
  "source": {
    "index": "cluster0:my-index-000001"
  },
  "dest": {
    "index": "my-index-000001"
  }
}
'
```

Closes #89120

### CODE REVIEW

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        final List<String> seedAddresses = seedNodes.stream().map(node -> {
+            final TransportService transportService = cluster(clusterAlias).getInstance(TransportService.class, node);
+            return transportService.boundAddress().publishAddress().toString();
+        }).toList();
+
+        LOGGER.info("--> use sniff mode with seed [{}], remote nodes [{}]", Collectors.joining(","), seedNodes);
+        settings.putNull(remoteClusterSettingPrefix + "proxy_address")
+            .put(remoteClusterSettingPrefix + "mode", "sniff")
+            .put(remoteClusterSettingPrefix + "seeds", String.join(",", seedAddresses))
+            .put(remoteClusterSettingPrefix + "skip_unavailable", true)
```
Comment (DaveCTurner) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:68]: I think we should support reindexing over CCS even with `skip_unavailable: false` (so that an unavailable remote causes the reindex to fail), but as implemented here that isn't the case.

**Review by DaveCTurner — DISMISSED** [review]
I'm not comfortable with quietly adding the `ignore_unavailable` search option like this. Often folks will want the reindex to fail if the source index is unavailable, but with this option I think it'll report success without having copied everything over. Also we need this to work on remote clusters with `skip_unavailable: false`.

**Comment by volodk85** [comment]
Thank you @DaveCTurner for having a look.

I added `ignore_unavailable` to Reindex API query parameters. Now in order to use reindex from remote, users have to specify it explicitly. `ignore_unavailable` index option does not depend on `skip_unavailable` remote cluster setting.
```
curl -X POST "localhost:19200/_reindex?ignore_unavailable=true" -H 'Content-Type: application/json' -d'
{
  "source": {
    "index": "cluster0:my-index-000001"
  },
  "dest": {
    "index": "my-index-000001"
  }
}
```

With this change Reindex API behaves as follow (assuming remote index does not exists):
`ignore_unavailable`: **true**
Local -> Local: Call successful 
Remote -> Local: Call sucessful
both report zeros in total/create/update/delete metrics
```
{"took":24,"timed_out":false,"total":0,"updated":0,"created":0,"deleted":0,"batches":0,"version_conflicts":0,"noops":0,"retries":{"bulk":0,"search":0},"throttled_millis":0,"requests_per_second":-1.0,"throttled_until_millis":0,"failures":[]}
```
`ignore_unavailable`: **false**
Local -> Local: Operation fails (index not found)
Remote -> Local: Operation fails due to gating in [IndexNameExpressionResolver.java](https://github.com/elastic/elasticsearch/blob/2e6834ad1103f7eab2e5a9cf2dcb5816ec4d2d17/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L1612-L1622) **Call to remote does not happen at all!**

>Often folks will want the reindex to fail if the source index is unavailable, but with this option I think it'll report success without having copied everything over

I agree that API should behave in the same way regardless we query local or remote index - if it does not exist, there should be clear error statement.

Now I wonder why the condition which mandates presence of `ignore_unavailable` flag for querying remote index in [IndexNameExpressionResolver.java](https://github.com/elastic/elasticsearch/blob/2e6834ad1103f7eab2e5a9cf2dcb5816ec4d2d17/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L1612-L1622) was ever introduced?
If it only for fixing datemath expression as commit message suggest (Fix indices resolver for datemath with colon  (#92973) Albert Zaharovits* 2023-01-18, 9:41 a.m.) IMHO it sounds like using wrong tool to solve the problem.

@albertzaharovits could you please comment?

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        assertAcked(client(REMOTE_CLUSTER).admin().indices().prepareCreate("source-index-001"));
+        final int docsNumber = indexDocs(client(REMOTE_CLUSTER), "source-index-001");
+
+        final String sourceIndexInRemote = REMOTE_CLUSTER + ":" + "source-index-001";
+
+        final IllegalArgumentException e = expectThrows(
+            IllegalArgumentException.class,
+            () -> new ReindexRequestBuilder(client(LOCAL_CLUSTER), ReindexAction.INSTANCE).source(sourceIndexInRemote)
+                .destination("desc-index-001")
+                .ignoreUnavailable(false)
```
Comment (DaveCTurner) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:150]: In contrast to this test, we want this case to succeed (i.e. not to throw an exception).

Comment (volodk85) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:150]: Exactly. I'm trying to sort it our with search team. Will keep you posted

File: docs/reference/docs/reindex.asciidoc [hunk 3]
```diff
@@ -489,6 +489,10 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards
 (Optional, Boolean) If `true`, the request blocks until the operation is complete.
 Defaults to `true`.
 
+`ignore_unavailable`::
+(Optional, Boolean) If false, the request returns an error if it targets a missing or closed index. Defaults to false.
+NOTE:: Should be set to `true` to perform reindexing from remote when index name is prefixed with remote cluster alias e.g. `cluster_name:index_name`
+
```
Comment (albertzaharovits) [inline docs/reference/docs/reindex.asciidoc:495]: I don't think this is a reasonable interface for the reindex API.

As it's possible to have a multiple reindex sources in the same reindex request, it could be that local and remote sources are mixed in the same request. In this case it's reasonable to support `ignore_unavailable=false` (which is the default), at least for the local indices (though I would prefer we also support it for remote indices too, which is how the search endpoint works, I think).

I noticed that the search request honours either `ignore_unvailable` options when using remote clusters (CCS).
Reindex uses search under the hood, so it should work, in principle... But, from what I can tell, it's the `ReindexValidator#validateAgainstAliases` check that trips it up. This check ought to prevent reindexing back onto the same index via an alias. I think it's this check that we need to adapt.

Is a remote index allowed as the **destination**? If not, and remote indices are only allowed as sources, then I think the easy fix is to exclude remote names from the list that the `validateAgainstAliases` check runs over, because it's not possible that a remote index name refer to a local index name.

Comment (volodk85) [reply docs/reference/docs/reindex.asciidoc]: No, seems remote index is **NOT** allowed as destination

**Comment by volodk85** [comment]
The core change here is to stop evaluating remote expressions on local node. If the expression is remote, it will be evaluated on remote node when request reaches target node. By doing this, it is possible to lift current restriction of having `ignore_unvailable` flag whenever we try to _reindex from remote.

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 4]
```diff
@@ -248,20 +248,21 @@ protected static Collection<String> resolveExpressions(Context context, String..
             if (expressions == null || expressions.length == 0 || expressions.length == 1 && Metadata.ALL.equals(expressions[0])) {
                 return List.of();
             } else {
-                return ExplicitResourceNameFilter.filterUnavailable(
-                    context,
-                    DateMathExpressionResolver.resolve(context, List.of(expressions))
-                );
+                // Do not evaluate remote expressions on local node, instead let remote node evaluate them when request arrives there
+                List<String> localExpressions = RemoteResourceNameFilter.filterRemoteIndexNames(List.of(expressions));
+                return ExplicitResourceNameFilter.filterUnavailable(context, DateMathExpressionResolver.resolve(context, localExpressions));
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:253]: I think the change to not "resolve" remote index names should be done in a caller to this method, see for example `TransportSearchAction#resolveLocalIndices`. This class is already very complex and brittle and I would suggest we avoid having it also deal with remote names, if possible (eg what if remote names are wildcards?).

**Review by albertzaharovits — CHANGES_REQUESTED** [review]

File: server/src/internalClusterTest/java/org/elasticsearch/get/GetActionIT.java [hunk 5]
```diff
@@ -824,11 +825,8 @@ void indexSingleDocumentWithStringFieldsGeneratedFromText(boolean stored, boolea
     }
 
     public void testGetRemoteIndex() {
-        IllegalArgumentException iae = expectThrows(IllegalArgumentException.class, () -> client().prepareGet("cluster:index", "id").get());
-        assertEquals(
-            "Cross-cluster calls are not supported in this context but remote indices were requested: [cluster:index]",
-            iae.getMessage()
-        );
+        IndexNotFoundException iae = expectThrows(IndexNotFoundException.class, () -> client().prepareGet("cluster:index", "id").get());
+        assertEquals("no such index [cluster:index]", iae.getMessage());
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/get/GetActionIT.java:829]: I suspect this is fundamentally the same as Albert's comment, but I don't think we should change the behaviour here. Cross-cluster GET doesn't really make sense, and it doesn't look like we're implementing cross-cluster GET anyway so it seems strange to report `no such index` here. What if the remote index _does_ exist?

**Comment by DaveCTurner** [comment]
Could we try the approach [mentioned in my comment on the issue](https://github.com/elastic/elasticsearch/issues/89120#issuecomment-1205256149) and just skip remote indices when validating the reindex request? In other words, adjust `org.elasticsearch.reindex.ReindexValidator#validateAgainstAliases` rather than the `IndexNameExpressionResolver` itself? We're only trying to avoid reindexing with the same source and target, but remote indices cannot be a target in any case so there's no need to resolve them here.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java (hunk 2 above)
Comment (volodk85) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:159]: this conversation was regarding test case
https://github.com/volodk85/elasticsearch/blob/2ae560e587506136e1dfc28f12124321f99b9c7d/modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java#L140-L166

**Comment by volodk85** [comment]
Recap of changes in this PR
1. `IndexNameExpressionResolver ` is not touched
2. reindex api now respects `ignore_unavailable` request option which is propagated in underlying search request
3. Fix for datemath expression: remote name with complex datemath expressions containing `:` symbol was treated as a "remote-of-remote" names

**Review by albertzaharovits — APPROVED** [review]
LGTM
Thank you for the iterations!

This API has a strangely particular design. My quibbly preference is that the  `ignore_unavailable` boolean parameter sits better under the "source" request body parameter (which also contains the "index" parameter). But this is just my slight preference and I almost never have used this API, nor have participated in its design, so you might get better opinions from somebody else.

File: docs/reference/docs/reindex.asciidoc [hunk 6]
```diff
@@ -489,6 +489,9 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=wait_for_active_shards
 (Optional, Boolean) If `true`, the request blocks until the operation is complete.
 Defaults to `true`.
 
+`ignore_unavailable`::
+(Optional, Boolean) If false, the request returns an error if it targets a missing or closed index. Defaults to false.
```
Comment (DaveCTurner) [inline docs/reference/docs/reindex.asciidoc:493]: Is this required to fix the issue? It's not a bad thing to add for sure but I'd prefer it to be a separate PR unless it's intimately linked to the bugfix.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasToString;
+
+public class CrossClusterReindexIT extends AbstractMultiClustersTestCase {
```
Comment (DaveCTurner) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:29]: Are there equivalent tests for within-cluster reindex actions? Including the date-math ones? I couldn't see any, but it'd be good to have some.

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java [hunk 8]
```diff
@@ -74,7 +75,10 @@ protected Map<String, List<String>> groupClusterIndices(Set<String> remoteCluste
         Map<String, List<String>> perClusterIndices = new HashMap<>();
         Set<String> clustersToRemove = new HashSet<>();
         for (String index : requestIndices) {
-            int i = index.indexOf(RemoteClusterService.REMOTE_CLUSTER_INDEX_SEPARATOR);
+            // ensure that `index` is a remote name and not a datemath expression which includes ':' symbol
+            // since datemath expression after evaluation should not contain ':' symbol
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java:79]: Hmm this raises questions about the interaction between cross-cluster search and date-math expressions that I do not feel qualified to answer. @javanna perhaps you have opinions about this? Are these things supposed to work together?

**Review by DaveCTurner — COMMENTED** [review]
My main comment is that you've found something surprising (to me) about the interaction of date-math expressions and cross-cluster search and that's very much in the domain of @elastic/es-search. I'm not sure these things are even supposed to work, and if they are I wonder about the test coverage because it looks like you're having to make some fairly far-reaching behaviour changes to get this to work.

File: modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java [hunk 9]
```diff
@@ -138,12 +139,32 @@ static void validateAgainstAliases(
...
+        );
+    }
+
+    private static boolean isRemoteExpression(String expression) {
+        // remote index name contains ':' symbol inside their names (remote cluster name separator), e.g. cluster0:index-name
+        // in the same time date-math `expression` can also contain ':' symbol inside its name
+        // to distinguish between those two, given `expression` is pre-evaluated using date-math resolver
+        // after evaluation date-math `expression` should not contain ':' symbol
+        // otherwise if `expression` is legit remote name, ':' symbol remains
+        return IndexNameExpressionResolver.resolveDateMathExpression(expression).contains(":");
```
[the remark this answers is not in the document]
Comment (volodk85) [reply modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java:168]: Reindex API does piggyback Search API which should support complex date-math index names ([doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/api-conventions.html#api-date-math-index-names)) but there is a bug there. Reindex API inherits the bug.

Since to decide if given index name is remote we lookup for semicolon, this code has to eval name in advance.

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java (hunk 8 above)
Comment (volodk85) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java:79]: Currently this code trips over if search is queried with complex date-math index name
```
curl -X GET "localhost:9200/%3Cdatemath-%7B2001-01-01-13%7C%7C%2B1h%2Fh%7Byyyy-MM-dd-HH%7C-07%3A00%7D%7D%3E/_count?error_trace"
```

{"error":{"root_cause":[{"type":"no_such_remote_cluster_exception","reason":"no such remote cluster: [<datemath-{2001-01-01-13||+1h/h{yyyy-MM-dd-HH|-07]","stack_trace":"org.elasticsearch.transport.NoSuchRemoteClusterException: no such remote cluster: [<datemath-{2001-01-01-13||+1h/h{yyyy-MM-dd-HH|-07]\n\tat org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.cluster.metadata.ClusterNameExpressionResolver.resolveClusterNames(ClusterNameExpressionResolver.java:41)\n\tat org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.transport.RemoteClusterAware.groupClusterIndices(RemoteClusterAware.java:85)\n\tat org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.transport.RemoteClusterService.groupIndices(RemoteClusterService.java:181)

File: docs/reference/docs/reindex.asciidoc (hunk 6 above)
Comment (volodk85) [reply docs/reference/docs/reindex.asciidoc]: I added it since cross-cluster ops piggy-back regular search API. It could be very useful in [Reindex from multiple sources](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#docs-reindex-from-multiple-sources) use case should an user wants to continue script execution opposing to current fail-fast behaviour. I I see this APIs alignment (with search) as added value.

But, OK I'm reducing scope of this PR

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java (hunk 7 above)
Comment (volodk85) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:27]: There are no high-level (IT) tests for this, but there are bunch of them on index name resolvers level

Comment (volodk85) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/CrossClusterReindexIT.java:27]: I extended test suite for within-cluster reindexing with test cases focused on date math index names

File: server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java (hunk 8 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java:79]: I am afraid that a local date math expression containing `:` will be mistaken for a cross-cluster search, like raised above. That happens even if you escape the `:` (just tested it). On the contrary, I'd expect a cross-cluster search request including a date math expressions with a `:` to work ok.

I would recommend opening a separate PR to fix this shortcoming. Strange that it has not been reported so far, it probably means that date math expression with a `:` in them are not widely used, as they effectively don't work in Elasticsearch for a while, at least since we removed support for indices containing `:` in their names.

Comment (volodk85) [reply server/src/main/java/org/elasticsearch/transport/RemoteClusterAware.java]: OK, I'm moving out logic which handles complex datemath expressions to separate PR

**Comment by volodk85** [comment]
Jenkins run elasticsearch-ci/bwc

**Review by DaveCTurner — APPROVED** [review]
Change & tests look great now, thanks @volodk85. Could we retitle this something like `Add support for reindex over CCS` before merging it tho? "Reindexing from remote" means [using the HTTP API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html#reindex-from-remote) and that's not what we're doing here.

File: modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java [hunk 10]
```diff
@@ -138,12 +139,27 @@ static void validateAgainstAliases(
...
     }
+
+    private static SearchRequest skipRemoteIndexNames(SearchRequest source) {
+        return new SearchRequest(source).indices(
+            Arrays.stream(source.indices()).filter(name -> isRemoteExpression(name) == false).toArray(String[]::new)
+        );
+    }
+
+    private static boolean isRemoteExpression(String expression) {
+        return expression.contains(":");
```
Comment (javanna) [inline modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java:163]: small nit: you could have used the existing RemoteClusterAware.REMOTE_CLUSTER_INDEX_SEPARATOR constant here.

Comment (DaveCTurner) [reply modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java:163]: TIL; @volodk85 would you make that change?

Comment (volodk85) [reply modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java:163]: If you don't mind I'll address this in follow up [PR](https://github.com/elastic/elasticsearch/pull/100646) (complex datemath things) which I opened yesterday for Search team.

Comment (DaveCTurner) [reply modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java:163]: Sounds good :+1:
